server.js

```diff
@@ -1,20 +1,119 @@
 const express = require("express");
 const cors = require("cors");
-
+const moment = require("moment"); // require
+moment().format("YYYY-MM-DD");
 const app = express();
 
 app.use(express.json());
 app.use(cors());
 
 //Use this array as your (in-memory) data store.
 const bookings = require("./bookings.json");
+const { body, validationResult } = require("express-validator");
 
 app.get("/", function (request, response) {
   response.send("Hotel booking server.  Ask for /bookings, etc.");
 });
 
 // TODO add your routes and helper functions here
+app.post(
+  "/bookings",
+  [body("email").isEmail().withMessage("Please input a valid email")],
+  function (request, response) {
+    const errors = validationResult(request);
+    if (!errors.isEmpty()) {
+      return response.status(422).json({ errors: errors.array() });
+    }
+    const expectedKeys = [
+      "title",
+      "firstName",
+      "surname",
+      "email",
+      "roomId",
+      "checkInDate",
+      "checkOutDate",
+    ];
+    for (let item in request.body) {
+      if (!request.body[item] || !expectedKeys.includes(item)) {
+        response
+          .status(400)
+          .json("Please ensure your are passing the correct fileds");
+        return;
+      }
+    }
+    if (!moment(request.body.checkOutDate).isAfter(request.body.checkInDate)) {
+      response.status(422).json("Checkout date must be after checkin date");
+      return;
+    }
+    let newBooking = {
+      id: bookings[bookings.length - 1].id + 1,
+      title: request.body.title,
+      firstName: request.body.firstName,
+      surname: request.body.surname,
+      email: request.body.email,
+      roomId: request.body.roomId,
+      checkInDate: request.body.checkInDate,
+      checkOutDate: request.body.checkOutDate,
+    };
+    bookings.push(newBooking);
+    response.json(bookings);
+  }
+);
+
+app.get("/bookings", function (request, response) {
+  response.send(bookings);
+});
+
+app.get("/bookings/search", function (request, response) {
+  let searchTerm = request.query.term;
+  let date = request.query.date;
+  if (moment(date, "YYYY-MM-DD", true).isValid()) {
+    let results = bookings.filter(
+      (booking) =>
+        date === booking.checkInDate ||
+        moment(date).isBetween(booking.checkInDate, booking.checkOutDate)
+    );
```

Comment:
Another good use of the moment library. It also has a neat way of including or excluding the start and end date of the range, using brackets - https://momentjscom.readthedocs.io/en/latest/moment/05-query/06-is-between/ 
If you use this, you don't have to do the first check - 
date === booking.checkInDate. 
That's more advanced than you need for this challenge, but it's worth having a look.

Reply:
I got this from the same page actually, but didn't pay attention to the inclusivity parameter, will implement it now. thanks for bringing this to my attention.